### DISCUSSION TITLE
Semantic Segmentation Task using custom dataset

### DISCUSSION THREAD

**Discussion by gtgrp-user** [torchgeo/torchgeo#1418] [Q&A] [answered]
Hi,

I have surface water dataset which I would like to use to train a semantic segmentation model. Both training and validation datasets are loaded in dataloaders. I would like to add a few indices and augmentation so I created the appropriate transforms object. 

Now, do I need a custom data module derived from LightningDataModule to run the SemanticSegmentationTask?

**Comment by isaaccorley** [comment]
Creating a `LightningDataModule` isn't required but we do have some useful features built into our datamodules. You can train with only dataloaders like so: 

```

model = SemanticSegmentationTask()

trainer = Trainer()
trainer.fit(model, train_dataloader, val_dataloader)
trainer.test(test_dataloader)
```

You can find more thorough documentation of this [here](https://lightning.ai/docs/pytorch/stable/common/trainer.html).

**Reply by gtgrp-user** [reply]
When attempting the method mentioned above, I encountered the following error.


```
ValueError: A frozen dataclass was passed to `apply_to_collection` but this is not allowed.
```

Code used:
```python
task = SemanticSegmentationTask(
    loss="jaccard",
    model="deeplabv3+",
    backbone="resnet50",
    weights='imagenet',
    in_channels=9,
    num_classes=2,
    learning_rate=0.1,
    learning_rate_schedule_patience=5,
    ignore_index=None
)
task.configure_optimizers()

accelerator = "gpu" if torch.cuda.is_available() else "cpu"
default_root_dir = os.path.join('./', "experiments")
checkpoint_callback = ModelCheckpoint(
    monitor="val_loss", dirpath=default_root_dir, save_top_k=1, save_last=True
)
early_stopping_callback = EarlyStopping(monitor="val_loss", min_delta=0.00, patience=10)
logger = TensorBoardLogger(save_dir=default_root_dir, name="tutorial_logs")

trainer = Trainer(
    accelerator=accelerator,
    callbacks=[checkpoint_callback, early_stopping_callback],
    enable_progress_bar = True,
    fast_dev_run=fast_dev_run,
    log_every_n_steps=1,
    logger=logger,
    min_epochs=1,
    max_epochs=max_epochs,
)

trainer.fit(task, train_dataloader, valid_dataloader)
```

**Reply by isaaccorley** [reply]
A quick google search found that this has to do with your [dataset containing a dataclass](https://github.com/Lightning-AI/lightning/issues/10873). Could you share the dataset code?

Also there is no need to call `task.configure_optimizers()`

**Reply by gtgrp-user** [reply]
Here is the datasets and dataloaders
```python
train_imgs = RasterDataset(root=(root/'tra_scene').as_posix(), crs='epsg:6933', res=10, transforms=transforms)
train_msks = RasterDataset(root=(root/'tra_truth').as_posix(), crs='epsg:6933', res=10)

valid_imgs = RasterDataset(root=(root/'val_scene').as_posix(), crs='epsg:6933', res=10, transforms=transforms)
valid_msks = RasterDataset(root=(root/'val_truth').as_posix(), crs='epsg:6933', res=10)

# IMPORTANT
train_msks.is_image = False
valid_msks.is_image = False

train_dset = train_imgs & train_msks
valid_dset = valid_imgs & valid_msks

# ROI to select training data from a specific regions 
train_sampler = RandomGeoSampler(train_imgs, size=512, length=260, units=Units.PIXELS)
valid_sampler = RandomGeoSampler(valid_imgs, size=512, length=128, units=Units.PIXELS)

train_dataloader = DataLoader(train_dset, sampler=train_sampler, batch_size=batch_size, collate_fn=stack_samples)
valid_dataloader = DataLoader(valid_dset, sampler=valid_sampler, batch_size=batch_size, collate_fn=stack_samples)
```

Another issue I have is adding a transform in the mask dataset. The AugmentationSequential allows data_keys ['image','mask'] together. So the best option would be to apply the transform after loading the datasets in a data loader. But I need to learn how to do it and whether it's the best practice!

**Reply by adamjstewart** [reply]
> ```
> ValueError: A frozen dataclass was passed to `apply_to_collection` but this is not allowed.
> ```

This _could_ be due to the non-Tensor values in each mini-batch (crs/bbox). See [here](https://github.com/microsoft/torchgeo/blob/v0.4.1/torchgeo/datamodules/geo.py#L265) for how we handle this in our GeoDataModule base class. It may be easiest just to make your own data module.

> Another issue I have is adding a transform in the mask dataset. The AugmentationSequential allows data_keys ['image','mask'] together. So the best option would be to apply the transform after loading the datasets in a data loader. But I need to learn how to do it and whether it's the best practice!

Yes, this is best practice. If you set `self.aug` (or `self.train_aug`, `self.val_aug`, etc. if you need them to differ) in your data module, they will be automatically applied to the entire mini-batch.

I would look at L7IrishDataModule and L8BiomeDataModule for similar example data modules. Let me know if you hit any snags.

**Reply by gtgrp-user** [reply]
Thanks, I will try this today.

**Reply by gtgrp-user** [reply]
I apologize for the long delay. I encountered a problem in my attempts to work with the L7IrishDataModule and L8BiomeDataModule, I encountered a problem. It appears that these modules are not available in the torchgeo version 0.4.1 that I have. As an alternative, I used the InriaDataModule, which is a NonGeoDataModule, and created my own data module. I was able to train a model successfully and managed to run the test and validate methods with the following metrics:
```
[{'val_loss': 0.08770599216222763,
  'val_MulticlassAccuracy': 0.9624531269073486,
  'val_MulticlassJaccardIndex': 0.9399352073669434}]

[{'test_loss': 0.0920742079615593,
  'test_MulticlassAccuracy': 0.9709367752075195,
  'test_MulticlassJaccardIndex': 0.9543617367744446}]
```
I am now able to perform predictions. However, I'm wondering if the torchgeo library also offers a method to perform inference on full-size Sentinel-2 images?

**Reply by adamjstewart** [reply, marked as answer]
The method to perform inference on full-size uncurated raster files would be to use a GeoDataModule instead of a NonGeoDataModule. [L7IrishDataModule](https://github.com/microsoft/torchgeo/blob/main/torchgeo/datamodules/l7irish.py) and [L8BiomeDataModule](https://github.com/microsoft/torchgeo/blob/main/torchgeo/datamodules/l8biome.py) are available in main if you want to take a look and use them for reference. [ChesapeakeCVPRDataModule](https://github.com/microsoft/torchgeo/blob/main/torchgeo/datamodules/chesapeake.py) and [NaipChesapeakeDataModule](https://github.com/microsoft/torchgeo/blob/main/torchgeo/datamodules/naip.py) are also GeoDataModules.

**Reply by geografif** [reply]
Hi, recently I tried to replicate the OP's workflow. Upon trying to fit the pytorch lightning trainer, it does not recognize `SemanticSegmentationTask` as a `LightningModule`. Is there any way to resolve this?

> `TypeError: `model` must be a `LightningModule` or `torch._dynamo.OptimizedModule`, got `SemanticSegmentationTask`

**Reply by adamjstewart** [reply]
Hi @geografif, can you share the code you are running that results in this error? A minimal reproducible example would help me debug this.

**Reply by isaaccorley** [reply]
@geografif this is probably because you are using `pytorch_lightning.Trainer`. You need to update to `lightning` and use `lightning.Trainer` instead.

**Reply by geografif** [reply]
Hi @isaaccorley @adamjstewart I used `lightning.Trainer`  and it worked. Thanks!

**Comment by MohajaneM** [comment]


**Reply by adamjstewart** [reply]
Are you using `GeoDataModule`? Can you share your dataset/datamodule code?

**Comment by MohajaneM** [comment]
Hello @adamjstewart  , I have fixed the error with GeoDataModule , thank you for your reply.


class CustomGeoDataModule(GeoDataModule):
    def __init__(self, dataset, label_polygons, width_sample, batch_size):
        super().__init__()
        self.dataset_train = dataset[0]
        self.dataset_val = dataset[1]
        self.label_polygons = label_polygons
        self.size = width_sample
        self.batch_size = batch_size

    def setup(self, stage: str = None):
        # Already assigned externally
        pass

    def train_dataloader(self):
        sampler = RandomGeoSamplerIntersectingPolygons(
            dataset=self.dataset_train,
            label_polygons=self.label_polygons,
            size=self.size,
            length=len(self.dataset_train),
            background_threshold=0.2,
            background_value=0,
        )
        return self._create_dataloader(self.dataset_train, sampler)

    def val_dataloader(self):
        sampler = RandomGeoSamplerIntersectingPolygons(
            dataset=self.dataset_val,
            label_polygons=self.label_polygons,
            size=self.size,
            length=len(self.dataset_val),
            background_threshold=0.2,
            background_value=0,
        )
        return self._create_dataloader(self.dataset_val, sampler)

**Reply by adamjstewart** [reply]
I think using a GeoDataModule will solve your problem. GeoDataset returns the CRS and bounding box of each image. However, PyTorch Lightning doesn't know how to transfer that to the GPU. GeoDataModule pops those keys so that PyTorch Lightning doesn't get confused. You could also customize the dataset to pop those keys as well.
